sumpy/fmm.py

```diff
@@ -483,6 +604,68 @@ def eval_direct(self, target_boxes, source_box_starts,
 
         return (pot, SumpyTimingFuture(self.queue, events))
 
+    @memoize_method
+    def multipole_to_local_precompute(self, src_rscale):
+        m2l_precomputed_exprs = self.m2l_precomputed_exprs_zeros()
+        events = []
+        for lev in range(self.tree.nlevels):
+            order = self.level_orders[lev]
+            precompute_kernel = \
+                self.code.m2l_optimized_precompute_kernel(order, order)
+
+            translation_classes_level_start, precomputed_exprs_view = \
+                    self.m2l_precomputed_exprs_view(m2l_precomputed_exprs, lev)
+
+            ntranslation_classes = precomputed_exprs_view.shape[0]
+
+            if ntranslation_classes == 0:
+                events.append([])
+                continue
+
+            m2l_translation_vectors = (
+                    self.translation_classes_data.m2l_translation_vectors())
+
+            evt, _ = precompute_kernel(
+                self.queue,
+                src_rscale=src_rscale,
+                translation_classes_level_start=translation_classes_level_start,
+                ntranslation_classes=ntranslation_classes,
+                m2l_precomputed_exprs=precomputed_exprs_view,
+                m2l_translation_vectors=m2l_translation_vectors,
+                ntranslation_vectors=m2l_translation_vectors.shape[1],
+                **self.kernel_extra_kwargs
+            )
+
+            events.append([evt])
+
+        return m2l_precomputed_exprs, events
+
+    def _use_multipole_to_local_precompute(self, kwargs_for_multipole_to_local,
+            lev):
+        """This method is used for addin the information needed for a
+        multipole-to-local translation with precomputation to the keywords
+        passed to multipole-to-local translation.
+        """
+        if not self.supports_optimized_m2l:
+            return
+        src_rscale = kwargs_for_multipole_to_local["src_rscale"]
+        m2l_precomputed_exprs, events = \
+            self.multipole_to_local_precompute(src_rscale)
+        translation_classes_level_start, precomputed_exprs_view = \
+            self.m2l_precomputed_exprs_view(m2l_precomputed_exprs, lev)
+        kwargs_for_multipole_to_local["m2l_precomputed_exprs"] = \
+                precomputed_exprs_view
+        kwargs_for_multipole_to_local["translation_classes_level_start"] = \
+                translation_classes_level_start
+        kwargs_for_multipole_to_local["m2l_translation_classes_lists"] = \
+            self.translation_classes_data.m2l_translation_classes_lists()
+
+        if "wait_for" in kwargs_for_multipole_to_local:
+            kwargs_for_multipole_to_local["wait_for"].append(events[lev])
+        else:
+            kwargs_for_multipole_to_local["wait_for"] = events[lev]
+        return
```

Comment:
```suggestion
```

Reply:
Fixed in a67dbed770e63bf23bd3dac891f337b2805a2ac3